commit 26e8c76ca84999fa5c0e46a9fc3aa7de80be2e9c
Author: Eelco Dolstra <edolstra@gmail.com>
Date:   Mon Oct 10 17:12:47 2022 +0200

    Return time_t in the Unix epoch

diff --git a/lib/zip_dirent.c b/lib/zip_dirent.c
index 7fd2f7ce..5c050b4c 100644
--- a/lib/zip_dirent.c
+++ b/lib/zip_dirent.c
@@ -1018,7 +1018,7 @@ _zip_d2u_time(zip_uint16_t dtime, zip_uint16_t ddate) {
     tm.tm_min = (dtime >> 5) & 63;
     tm.tm_sec = (dtime << 1) & 62;
 
-    return mktime(&tm);
+    return timegm(&tm);
 }
 
 
